docs/content/posts/2024-11-26-introducing-new-prometheus-connector.md

```diff
@@ -0,0 +1,202 @@
+  int64 timestamp = 2;
+}
+```
+
+The POST payload is serialized as Protobuf 3, compressed with [Snappy](https://github.com/google/snappy).
+
+### Specifications constraints
+
+Specifications do not impose any constraints on repeating `TimeSeries` with the same set of `Label`, but it does require strict ordering of `Sample` belonging to the same time-series (by `timestamp`) and `Label` in a single `TimeSeries` (by name).
+
+{{< hint info >}}
+The term "time-series" is overloaded, referring to both:
+1. A unique series of samples in the datastore, identified by unique set of labels,
+    and
+2. `TimeSeries` as a block of the `WriteRequest`.
+
+The two concepts are obviously related, but a `WriteRequest` may contain multiple `TimeSeries` elements referring to the same datastore *time-series*.
+{{< /hint >}}
+
+### Prometheus design principles vs Flink design principles
+
+Flink is designed with a consistency-first approach. By default, any unexpected error causes the streaming job to fail and restart from the last checkpoint.
+
+In contrast, Prometheus is designed with an availability-first approach, prioritizing fast ingestion over strict consistency. When a write request contains malformed entries, the entire request must be discarded and not retried. Additionally, samples belonging to the same time series (with the same dimensions or labels) must be written in strict timestamp order.
+
+You may have already spotted the issue: any malformed or out-of-order samples can act as a “poison pill” unless you drop the offending request and proceed.
+
+### Strict specifications, lenient implementations
+
+The [Remote-Write specifications](https://prometheus.io/docs/specs/remote_write_spec/) are very strict regarding the [ordering of `Labels`](https://prometheus.io/docs/specs/remote_write_spec/#labels) within a request, the [ordering of `Samples`](https://prometheus.io/docs/specs/remote_write_spec/#ordering) belonging to the same time-series, and preventing from [retrying rejected requests](https://prometheus.io/docs/specs/remote_write_spec/#retries-backoff).
+
+However, many Prometheus backend implementations are more lenient. They may support a limited, configurable out-of-order time window and may not enforce strict label ordering as long as other rules are followed.
+
+As a result, a connector that strictly enforces the Remote-Write specifications may limit the user’s flexibility compared to what their specific Prometheus implementation allows.
+
+### Error-handling and guarantees
+
+As we have seen, the Remote-Write specifications require that certain rejected writes—specifically those related to malformed and out-of-order samples (typically 400 responses)—must not be retried. The connector has no choice but to discard the entire offending request and continue with the next record. The specifications do not define a standard response format, and there is no reliable way to automatically identify and selectively discard the offending samples.
```

Comment:
Would the connector send out of order samples? Or is that determined by the user records?

Reply:
The connector does not generate out-of-orderness itself.
For parallelism, we provide a KeySelector to partition the stream of input objects so that the same time-series are written by the same subtask. 
Upstream of that, it's responsibility of the user to correctly partition data to prevent accidental out-of-orderness. We made it very clear in the docs as well

Reply:
Let me clarify this